typhos/alarm.py

```diff
@@ -0,0 +1,380 @@
+    """
+    QtCore.Q_ENUMS(KindLevel)
+    QtCore.Q_ENUMS(AlarmLevel)
+    KindLevel = KindLevel
+    AlarmLevel = AlarmLevel
+
+    alarm_changed = QtCore.Signal(AlarmLevel)
+
+    def __init__(self, *args, **kwargs):
+        self._kind_level = KindLevel.HINTED
+        super().__init__(*args, **kwargs)
+        self.init_alarm_state()
+        self.alarm_changed.connect(self.set_alarm_color)
+
+    @QtCore.Property(KindLevel)
+    def kindLevel(self):
+        """
+        Determines which signals to include in the alarm summary.
+
+        If this is "hinted", only include hinted signals.
+        If this is "normal", include normal and hinted signals.
+        If this is "config", include everything except for omitted signals
+        If this is "omitted", include all signals
+        """
+        return self._kind_level
+
+    @kindLevel.setter
+    def kindLevel(self, kind_level):
+        # We must update the alarm config to add/remove PVs as appropriate.
+        self._kind_level = kind_level
+        self.update_alarm_config()
+
+    @QtCore.Property(str)
+    def channel(self):
+        """
+        The channel address to use for this widget.
+
+        If this is a happi:// channel, we'll create the device and
+        add it to this widget.
+
+        If this is a ca:// channel, we'll connect to the PV and include its
+        alarm information in the evaluation of this widget.
+
+        There is an assumption that you'll either be using this via one of the
+        channel options or by using "add_device" one or more times. There may
+        be some strange behavior if you try to set up this widget using both
+        approaches at the same time.
+        """
+        if self._channel:
+            return str(self._channel)
+        return None
+
+    @channel.setter
+    def channel(self, value):
+        if self._channel != value:
+            # Remove old connection
+            if self._channels:
+                self._channels.clear()
+                for channel in self._channels:
+                    if hasattr(channel, 'disconnect'):
+                        channel.disconnect()
+            # Load new channel
+            self._channel = str(value)
+            channel = HappiChannel(
+                address=self._channel,
+                tx_slot=self._tx,
+                connection_slot=partial(self.update_connection,
+                                        addr=self._channel),
+                severity_slot=partial(self.update_severity,
+                                      addr=self._channel),
+                )
+            self._channels = [channel]
+            # Connect the channel to the HappiPlugin
+            if hasattr(channel, 'connect'):
+                channel.connect()
+
+    def _tx(self, value):
+        """Receive information from happi channel"""
+        self.add_device(value['obj'])
+
+    def init_alarm_state(self):
```

Comment:
So nitpicky I am regretting typing this, but perhaps `reset_alarm_state` instead?

Reply:
true